docs/design/authorization.md

```diff
@@ -0,0 +1,21 @@
+# Authorization
+
+## Problem statement
+
+Smith typically has a very broad set of permissions because it needs to be able to manage objects of various kinds
+across several/many namespaces in a cluster. If permissions checks are not done properly, a malicious user may have
+a way to escalate their privileges.
+This document describes a way to address the problem of blocking ways to permissions escalation.
+
+## Proposed solution
+
+The idea is to capture identity of the user who creates/updates a `Bundle` and impersonate them when working with
+objects of that `Bundle`. That way Smith becomes just a tool that automates operations of behalf of the user rather
+than a more privileged actor that can do something the user cannot.
+
+Implementation is straightforward.
+
+1. Use a [MutatingAdmissionWebhook](https://kubernetes.io/docs/admin/admission-controllers/#mutatingadmissionwebhook-beta-in-19)
+to capture information about the user's identity as a field in the `Bundle`.
+2. [Impersonate the user](https://kubernetes.io/docs/admin/authentication/#user-impersonation) when making any requests
```

Comment:
Note for myself and other reviewers:
> To impersonate a user, group, or set extra fields, the impersonating user must have the ability to perform the “impersonate” verb on the kind of attribute being impersonated (“user”, “group”, etc.). For clusters that enable the RBAC authorization plugin, the following ClusterRole encompasses the rules needed to set user and group impersonation headers:
```
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: impersonator
rules:
- apiGroups: [""]
  resources: ["users", "groups", "serviceaccounts"]
  verbs: ["impersonate"]
```
i.e. Smith will need to have this rule to be able to impersonate users for a given list of resources.